CSharpSourceCode/CampaignSupport/WanderersController.cs

```diff
@@ -0,0 +1,85 @@
+﻿using HarmonyLib;
+using System.Collections.Generic;
+using System.Linq;
+using TaleWorlds.CampaignSystem;
+using TaleWorlds.CampaignSystem.Actions;
+using TaleWorlds.CampaignSystem.SandBox.CampaignBehaviors;
+using TaleWorlds.Core;
+using TOW_Core.Utilities.Extensions;
+
+namespace TOW_Core.CampaignSupport
+{
+    public class WanderersController : CampaignBehaviorBase
+    {
+        private List<Hero> _companions;
+        private Dictionary<Settlement, CampaignTime> _companionSettlements;
+
+        public override void SyncData(IDataStore dataStore)
+        {
+        }
+
+        public override void RegisterEvents()
+        {
+            CampaignEvents.OnAfterSessionLaunchedEvent.AddNonSerializedListener(this, OnAfterSessionLaunched);
+            CampaignEvents.AfterSettlementEntered.AddNonSerializedListener(this, OnAfterSettlementEntered);
+        }
+
+        private void OnAfterSessionLaunched(CampaignGameStarter gameStarter)
+        {
+            var urbanBehavior = Campaign.Current.GetCampaignBehavior<UrbanCharactersCampaignBehavior>();
+            if (urbanBehavior != null)
+            {
+                _companions = Traverse.Create(urbanBehavior).Field("_companions").GetValue<List<Hero>>();
+                _companionSettlements = Traverse.Create(urbanBehavior).Field("_companionSettlements").GetValue<Dictionary<Settlement, CampaignTime>>();
+            }
+        }
+
+        private void OnAfterSettlementEntered(MobileParty mobileParty, Settlement settlement, Hero hero)
+        {
+            if (hero == null || !hero.IsHumanPlayerCharacter)
+            {
+                return;
+            }
+
+            var wanderer = Hero.AllAliveHeroes.FirstOrDefault(h => h.IsWanderer && h.CurrentSettlement == settlement);
```

Comment:
This just checks the first wanderer in the current settlement. There can be more than one, so this should be a list of wanderers, not FirstOrDefault().